### PR TITLE
split isolation pkg to implement concurrency traffic control

### PR DESCRIPTION
**Pull request by louyuting**
### Describe what this PR does / why we need it
split isolation pkg to implement concurrency traffic control

### Special notes for reviews

### CODE REVIEW

File: core/isolation/rule_manager.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	ruleMap = m
+	return
+}
+
+func ClearRules() error {
+	_, err := LoadRules(nil)
+	return err
+}
+
+func GetRules() []Rule {
```
Comment (louyuting) [inline core/isolation/rule_manager.go:50]: I  will add comments later

Comment (louyuting) [reply core/isolation/rule_manager.go:50]: Updated

File: core/isolation/slot.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+			} else {
+				curCount = 0
+				logging.Error(errors.New("negative concurrency"), "", "rule", rule)
+			}
+			if curCount+acquireCount > threshold {
+				return false, rule, curCount
+			}
+		}
+	}
+	return true, nil, curCount
```
Comment (sczyh30) [inline core/isolation/slot.go:46]: Maybe we could also carry the rule in normal result, or add comments about it (nil when passed).

File: core/isolation/rule_manager.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+// getRulesOfResource is an internal interface.
+func getRulesOfResource(res string) []*Rule {
+	rwMux.RLock()
+	defer rwMux.RUnlock()
+
+	resRules, exist := ruleMap[res]
+	if !exist {
+		return nil
+	}
+	ret := make([]*Rule, 0, len(resRules))
```
Comment (sanxun0325) [inline core/isolation/rule_manager.go:93]: Using copy Func may be more concise

File: core/isolation/rule_manager.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+var (
+	ruleMap = make(map[string][]*Rule)
+	rwMux   = &sync.RWMutex{}
+)
+
+// LoadRules loads the given isolation rules to the rule manager, while all previous rules will be replaced.
+func LoadRules(rules []*Rule) (updated bool, err error) {
+	updated = true
+	err = nil
```
[the remark this answers is not in the document]
Comment (louyuting) [reply core/isolation/rule_manager.go:20]: It just keep API consistency with other module

File: core/isolation/rule_manager.go (hunk 3 above)
Comment (louyuting) [reply core/isolation/rule_manager.go:93]: Updated

File: core/isolation/slot.go (hunk 2 above)
Comment (louyuting) [reply core/isolation/slot.go:46]: Comments had been added

**Review by sczyh30 — APPROVED** [review]
LGTM

**Pull request merged**